### ISSUE TITLE
master branch build failure on Apple M1

### ISSUE THREAD

**Issue by min-mwei** [ClickHouse/ClickHouse#27593]
It seems some recent change in cpu feature config, caused the build from master to fail on OSX/Apple M1 

cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo  ..
[ 97%] Linking CXX executable unit_tests_dbms
Undefined symbols for architecture arm64:
  "_arm_cpu_has_crc32", referenced from:
      _insert_string_stub in libz.a(functable.c.o)
      _quick_insert_string_stub in libz.a(functable.c.o)
      _crc32_stub in libz.a(functable.c.o)
  "_arm_cpu_has_neon", referenced from:
      _slide_hash_stub in libz.a(functable.c.o)
      _adler32_stub in libz.a(functable.c.o)
      _chunksize_stub in libz.a(functable.c.o)
      _chunkcopy_stub in libz.a(functable.c.o)
      _chunkcopy_safe_stub in libz.a(functable.c.o)
      _chunkunroll_stub in libz.a(functable.c.o)
      _chunkmemset_stub in libz.a(functable.c.o)
      ...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

**Comment by KochetovNicolai** [comment]
I didn't check for M1, but my mac 16 successfully build unit_tests_dbms from master:
```
/Applications/CLion.app/Contents/bin/cmake/mac/bin/cmake --build /Users/nik-kochetov/dev/ClickHouse/cmake-build-relwithdebinfo --target unit_tests_dbms -- -j 9
...
[100%] Linking CXX executable unit_tests_dbms
[100%] Built target unit_tests_dbms
```
Maybe submodules need to be updated? Try `git submodule update --init --recursive`

@kitaisreal can you please also check?

**Comment by min-mwei** [comment]
Yes, I think some recent changes in zlib-ng broke the M1 build. On x64 it's fine.

**Comment by oliveagle** [comment]
This is how I build Clickhouse on Apple M1.  

commit I'm building: `40f5e06a8d9b2074b5985a8042f3ebf9940c77f4`

Three problems: 
1. `zlib-ng` is breaking the compiling process. 
2. `std::filesystem:: weakly_canonical` keeps crashing the app.
3. minor problem `icu4c` always not found with fixed `-L/usr/local/opt/icu4c/lib`

##  `zlib-ng` 

two problems found with `zlib-ng`:

1. I have to switch to the official repository.  both `2.0.5` or `develop` can work.
2. Have to explicit set arch to `aarch64` in `detect-arch.cmake` file. 

```cmake
else()
    # Let preprocessor parse archdetect.c and raise an error containing the arch identifier
    set(ARCH "aarch64") # <-- hardcoded here.
    enable_language(C)
    try_run(
        run_result_unused
        compile_result_unused
        ${CMAKE_CURRENT_BINARY_DIR}
        ${CMAKE_CURRENT_LIST_DIR}/detect-arch.c
        COMPILE_OUTPUT_VARIABLE RAWOUTPUT
        CMAKE_FLAGS CMAKE_OSX_ARCHITECTURES=${CMAKE_OSX_ARCHITECTURES}
    )
```


## `std::filesystem:: weakly_canonical` problem

After the `zlib-ng` problem is solved.  CMake build will be finished without error. but generated binary won't start.  always crashed with SIGABRT.   see traces below. 

![image](https://user-images.githubusercontent.com/2270205/130060389-90b5fda0-e3ea-4c09-aea7-1a022b9c9776.png)

evidence pointing to`fs::weakly_canonical` in `DateLUT.cpp`.   I comment lines that are using `weakly_canonical` functions, it stops crashing anymore. 

then I replaced `std::filesystem` with `boost::filestatem` it also works.

```c++
#include <boost/filesystem.hpp>

std::string determineDefaultTimeZone()
{
    // namespace fs = std::filesystem;
    namespace fs = boost::filesystem;
```
![image](https://user-images.githubusercontent.com/2270205/130064851-bf753111-b94e-4179-9484-221da6b34fb0.png)

## `icu4c` problem

This problem won't crash the application nor break the compiling process though.

Homebrew on m1 is located under `/opt/homebrew`, so all libraries are installed there.  when building Clickhouse, all libraries are found with prefixes `/opt/homebrew` with the exception of `icu4c`. which keeps using `-L/usr/local/opt/icu4c/lib`.  I tried to hard code the path of the library in many different ways. but none of my attempts worked.  very odd. 

Finally, I have to make a symbolic link to fix it.

**Comment by armenzg** [comment]
Any update if anyone will be working on this in the near future? Thank you all for your hard work
